src/coreclr/System.Private.CoreLib/src/System/Reflection/ModifiedType.CoreCLR.cs

```diff
@@ -0,0 +1,50 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace System.Reflection
+{
+    internal partial class ModifiedType
+    {
+        private Signature? _signature; // Signature is a CoreClr-specific class.
+
+        /// <summary>
+        /// Called from FieldInfo, PropertyInfo and ParameterInfo to create a modified type tree.
+        /// </summary>
+        public static ModifiedType Create(Type unmodifiedType, int rootSignatureParameterIndex, Signature? signature)
+        {
+            ModifiedType modifiedType = Create(unmodifiedType, rootSignatureParameterIndex);
+            modifiedType._signature = signature;
+            return modifiedType;
+        }
+
+        public Signature? GetSignature() => Root._signature;
```

Comment:
This means that `_signature` is unused field in non-Root ModifiedTypes.

Can we avoid storing Root type in the ModifiedType and just store the Signature and indices into the signature? It will make the ModifiedType a bit smaller.

Reply:
> This means that _signature is unused field in non-Root ModifiedTypes.

Yes `_signature` is not used for non-roots. This approach is trying to allow re-use of the managed code across different runtimes. If I add the use of generics to include a `<TSignatureProvider>` (or use `object` instead of generics) which for CoreClr would be `Signature` then I can remove the need for `root` and have every node basically reference `signature`. This would help a bit for GC since there wouldn't be any cycles.

> Can we avoid storing Root type in the ModifiedType and just store the Signature and indices into the signature? It will make the ModifiedType a bit smaller.

Since each node is essentially stand-alone and doesn't know about its hierarchy (except for the root, which per above can be removed) each node needs all of this information. The same signature is used for all nodes in a given tree, so it can't store those. Note that we could combine all 3 integers into 1 which would mean "node index" but that would hurt understandability I think - today there are 3 levels of parsing in the native code and we'd need to combine them.

Reply:
I do not think we should be introducing extra complexity and inefficiencies just to make some more code shareable between the runtimes. I would be fine with sharing less code if it makes things simpler.

The ModifiedType is NYI for native AOT and Mono flavors, so it is hard to see why this is helping.

Reply:
I did remove the "Root" back reference and now each node is completely stand-alone which I think makes it simpler to understand since the lookup information is one location with no need to go back to the root.

> The ModifiedType is NYI for native AOT and Mono flavors, so it is hard to see why this is helping.

I think the latest approach doesn't really add much complexity.

Reply:
If NativeAot\Mono can't re-use the implementation, then refactoring back to just Core will be trivial: combing partial files and change the type of the "SignatureProvider" field from `object` to `Signature`. i.e. it's not really sacrificing anything currently.

Reply:
Notes on the design:
- Uses the 3 `int`s described in the code to find any node to get its signature (also described in the code):
  - The 1st `int` gets the root node from the field\parameter\property
  - The 2nd `int` is for recursion to skip past any intermediate function pointer depth levels. Note that any arrays, pointers or generic type parameters are also skipped but those don't increment this `int` since this PR doesn't support modifiers for them.
  - The 3rd `int` gets the appropriate node from the function pointer

Alternatives in NativeAot\Mono that would not be compatible:
- Cache and\or lookup various offsets (to each signature or to each node).
- Store the modifiers on a type handle (may have a special "modified type handle"). This was done in a prior prototype, but it added unnecessary runtime overhead\complexity.
- ...